dev/integration_tests/windows_startup_test/lib/main.dart

```diff
@@ -70,25 +70,41 @@ void main() async {
       throw 'Window should be hidden at startup';
     }
 
-    bool firstFrame = true;
-    ui.PlatformDispatcher.instance.onBeginFrame = (Duration duration) async {
-      if (await isWindowVisible()) {
-        if (firstFrame) {
-          throw 'Window should be hidden on first frame';
+    const int totalTestFrames = 10;
+    int frameCount = 0;
+    ui.PlatformDispatcher.instance.onBeginFrame = (Duration duration) {
+      // Our goal is to verify that it's `drawHelloWorld` that makes the window
+      // appear, not anything else. This requires checking the visibility right
+      // before drawing, but since `isWindowVisible` has to be async, and
+      // `FlutterView.render` (in `drawHelloWorld`) forbids async before it,
+      // this can not be done during a single onBeginFrame. However, we can
+      // verify in two frames to indirectly prove it: The first frame checks
+      // isWindowVisible(), while the subsequent frames draw. This ensures that
+      // no other mechanism can affect isWindowVisible in the first frame at all.
+      frameCount += 1;
+      // The window is made appear in the second frame.
+      // TODO(dkwingsmt): It should be frameCount > 2. Not sure why the window
+      // is still hidden at the beginning of the 3rd frame.
+      // https://github.com/flutter/flutter/issues/134986
+      final bool shouldBeVisible = frameCount > 3;
```

Comment:
There's lots of thread hopping happening here to show the window (ui thread -> raster thread -> platform thread -> UI thread), I don't think there's anything guaranteeing the window will be shown by the third frame. This might also change due to implementation details too.

I think the test would be better if implemented like:

```dart
void main() async {
  // ...

  if (await isWindowVisible()) {
    const String error = 'Window should be hidden at startup';
    visibilityCompleter.completeError(AssertionError(error));
    throw error;
  }

  bool firstFrame = true;
  ui.PlatformDispatcher.instance.onBeginFrame = (Duration duration) {
    try {
      if (firstFrame) {
        isWindowVisible().then((bool visible) {
          if (visible) {
            throw 'Window should be hidden on first frame';
          }

          // Rendering must happen within the frame lifecycle. Since checking
          // the window visibility is an asynchronous operation, we must
          // re-enter the frame lifecycle by scheduling another frame.
          firstFrame = false;
          ui.PlatformDispatcher.instance.scheduleFrame();
        });
      } else {
        // Draw something to trigger the callback that displays the window.
        drawHelloWorld(view);

        // Check the window's visibility in the background...
        unawaited(waitUntilWindowVisible(visibilityCompleter));
      }
    } catch (e, s) {
      visibilityCompleter.completeError(e, s);
      rethrow;
    }
  };
}

Future<void> waitUntilWindowVisible(Completer<String> visibilityCompleter) async {
  try {
    while (!await isWindowVisible()) {
      await Future<void>.delayed(const Duration(milliseconds: 100));
    }

    visibilityCompleter.complete('success');
  } catch (e, s) {
    visibilityCompleter.completeError(e, s);
    rethrow;
  }
}
```